### PR TITLE
[SPARK-44398][CONNECT] Scala foreachBatch API

### PR DESCRIPTION
**Pull request by rangadi**
This implements Scala foreachBatch(). The implementation basic and needs some more enhancements. The server side will be shared by Python implementation as well.

One notable hack in this PR is that it runs user's `foreachBatch()` with regular(legacy) DataFrame, rather than setting up remote Spark connect session and connect DataFrame. 

### Why are the changes needed?
Adds foreachBatch() support in Scala Spark Connect.

### Does this PR introduce _any_ user-facing change?
Yes. Adds foreachBatch() API

### How was this patch tested?
- A simple unit test.

### CODE REVIEW

**Comment by rangadi** [comment]
cc: @bogao007

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingForeachBatchHelper.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        sessionHolder.removeCachedDataFrame(dfId)
+      }
+    }
+  }
+
+  /**
+   * Handles setting up Scala remote session and other Spark Connect environment and then
+   * runs the provided foreachBatch function `fn`.
+   *
+   * HACK ALERT: This version does not atually set up Spark connect. Directly passes the DataFrame,
```
Comment (bogao007) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingForeachBatchHelper.scala:59]: I have 2 major questions regarding this:
- Is the missing part about setting up a Spark Connect session and converting the legacy DataFrame to Spark Connect DataFrame and being executed inside the Spark Connect session? Do we have any Scala example on setting up Spark Connect session on server side and use it?
- When is `getDataFrameOrThrow()` being called? Is it only needed for Python or do we need to get the DataFrame by ID inside the Spark Connect session for Scala?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 2]
```diff
@@ -292,3 +310,14 @@ class TestForeachWriter[T] extends ForeachWriter[T] {
...
+object StreamingQuerySuite {
+
+  // A foreach batch function for use in test above. Stores the count of records in a temp view.
+  def foreachBatchFnForTest1(df: DataFrame, batchId: Long): Unit = {
+    val count = df.count()
+    df.sparkSession
+      .createDataFrame(Seq((batchId, count)))
+      .createOrReplaceTempView("foreachBatchTest1")
+  }
+}
```
Comment (bogao007) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:323]: Nit: missing an empty line at the end of file

**Review by bogao007 — APPROVED** [review]
LGTM overall, left some questions regarding to the hack.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingForeachBatchHelper.scala (hunk 1 above)
Comment (rangadi) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingForeachBatchHelper.scala:59]: * Yes, it is about setting up spark remote session. I don't think there are examples of doing that in Scala. 
 * Not sure about the second one. Usually `df.sparkSession` gives the access to session.

**Comment by rangadi** [comment]
@zhenlineo could you review this? I am going to get the test working right.
We want to merge this ASAP as we are getting multiple streams of work done by 3.5 branch cut.

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 3]
```diff
@@ -215,15 +215,24 @@ message WriteStreamOperationStart {
   }
 
   StreamingForeachWriter foreach_writer = 13;
+  StreamingForeachFunction foreach_batch = 14;
```
Comment (zhenlineo) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:218]: I think you can rename the protos if it has not yet be released.

**Review by zhenlineo — APPROVED** [review]

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto (hunk 3 above)
Comment (rangadi) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:218]: Nice. Will change in this.

Comment (rangadi) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:218]: Replaced.

**Review by xinrong-meng — APPROVED** [review]

**Comment by xinrong-meng** [comment]
Merged to master, thanks!

**Pull request closed without merging**